### ISSUE TITLE
Problem with knitting documents when Rmd is in a readonly directory

### ISSUE THREAD

**Issue by anandkkumar** [rstudio/rmarkdown#1839]
Hi, 

We package Rmd files with our shared packages and in downstream apps we load and render them. So the source directory with the Rmd file is read-only (since the shared package is in a system library) and so we cannot knit directly in it. I tried to use the parameters intermediates_dir and knit_root_dir and set them to a custom output directory (which is writable) but I get an error when knitting the document.
```
!!! Error: Cannot write tmp file `tmp-pdfcrop-2449.tex' (Permission denied)!
Error: LaTeX failed to compile /home/ak6/output/TestDoc.tex. See https://yihui.org/tinytex/r/#debugging for debugging tips.
```

I looked through the [source code](https://github.com/rstudio/rmarkdown/blob/master/R/render.R) and it appears to be in these lines. It is explicitly setting the working directory to the dirname of the `input file`. 

```
  # execute within the input file's directory
  oldwd <- setwd(dirname(abs_path(input)))
  on.exit(setwd(oldwd), add = TRUE)
```

Can it use the `intermediates_dir` or `knit_root_dir` instead? 

The actual error shows up later in the call to `knitr::knit()`  since that calls getwd() but the lines above seem to be where it's changing the working directory. I also tried setting `root.dir` in knitr options as suggested [here](https://github.com/yihui/knitr/issues/277) but that didn't seem to work either. The only hack that I can think of is to temporarily copy the Rmd file from its read-only input location to `/tmp` or something like that, generate the output and then delete that temp file. Seems like a waste.

I don't know if this can be fixed based on how knitr works but thought I would ask. 

A reproducible example is below.

1) create directories called 'Rmd' and 'output'
2) rename the attached file to TestDoc.Rmd and copy to the 'Rmd' folder
3) make the 'Rmd' folder read-only
4) run the following code

```
rmarkdown::render(input="~/Rmd/TestDoc.Rmd",
                  output_format="pdf_document",
                  output_dir="~/output",
                  intermediates_dir="~/output",
                  knit_root_dir = "~/output",
                  quiet=TRUE)
```

[TestDoc.txt](https://github.com/rstudio/rmarkdown/files/4732514/TestDoc.txt)

Session info:
```
> xfun::session_info()
R version 3.6.3 (2020-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.4 LTS, RStudio 1.2.5042.1

Locale:
  LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
  LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
  LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
  LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

Package version:
  base64enc_0.1.3 compiler_3.6.3  digest_0.6.25   evaluate_0.14   glue_1.4.0     
  graphics_3.6.3  grDevices_3.6.3 highr_0.8       htmltools_0.4.0 jsonlite_1.6.1 
  knitr_1.28      magrittr_1.5    markdown_1.1    methods_3.6.3   mime_0.9       
  Rcpp_1.0.4.6    rlang_0.4.5     rmarkdown_2.1   stats_3.6.3     stringi_1.4.6  
  stringr_1.4.0   tinytex_0.22    tools_3.6.3     utils_3.6.3     xfun_0.13      
  yaml_2.2.1  
```

---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by cderv** [comment]
> The only hack that I can think of is to temporarily copy the Rmd file from its read-only input location to /tmp or something like that, generate the output and then delete that temp file. Seems like a waste.

I don't think this is a waste. `knitr` and `rmarkdown` work well when rendering a Rmd file from its directory. Doing differently must lead to deal with several paths stuff and it is often not as easy as a `file.copy` from the input file to a clean location where you have full rights.

Anyway, it is worth looking if anything could be improved. Thanks!
